File: code/java/start/src/ShopCart.java
import java.util.HashMap;
import java.util.Map;

public class ShopCart {
    private final Map<String, Integer> carts;

    public ShopCart() {
        carts = new HashMap<>();
    }

    public int size() {
        return carts.size();
    }

    public void addBook(String name) {
        int amount = carts.getOrDefault(name, 0) + 1;
        carts.put(name, amount);
    }

    public void addBook(String name, int num) {
        assert num > 0;
        int amount = carts.getOrDefault(name, 0) + num;
        carts.put(name, amount);
    }

    public int getAmount(String name) {
        return carts.getOrDefault(name, 0);
    }

    public static void main(String[] args) {
        ShopCart shopCart = new ShopCart();
        shopCart.addBook("8899");

        assert shopCart.size() == 1;
        assert shopCart.getAmount("8899") == 1;
        assert shopCart.getAmount("7788") == 0;

        shopCart.addBook("7788", 2);
        assert shopCart.getAmount("7788") == 2;

        shopCart.addBook("7788");
        assert shopCart.getAmount("7788") == 3;

        assert shopCart.size() == 2;
    }
}

File: code/python/start/shop_cart.py
class ShopCart:
    def __init__(self):
        self.carts = {}

    def size(self):
        return len(self.carts)

    def add_book(self, name, v=1):
        amount = self.carts.get(name, 0) + v
        self.carts[name] = amount

    def get_amount(self, name):
        return self.carts.get(name, 0)


if __name__ == '__main__':
    shop_cart = ShopCart()
    shop_cart.add_book("8899")

    assert shop_cart.size() == 1
    assert shop_cart.get_amount('8899') == 1
    assert shop_cart.get_amount('7788') == 0

    shop_cart.add_book('7788', 2)
    assert shop_cart.get_amount('7788') == 2

    shop_cart.add_book('7788')
    assert shop_cart.get_amount('7788') == 3

    assert shop_cart.size() == 2
